### PR TITLE
Petfinder

### PR DESCRIPTION
**Pull request by emcapone**
This is the Petfinder API implementation. I also created the basic service for bookmarks. There is no component displaying any bookmark data yet. I added a token refresh in the last commit.

I haven't worked with XML escaped characters before, so I am unsure if there's a better solution available than manually checking everything like I did in the PetfinderPet component. If there is a better solution please do let me know.

My next goals are to create the bookmark component and tackle bugs/features I previously skipped.

### CODE REVIEW

**Comment by emcapone** [comment]
image
image
image

File: src/app/petfinder/components/petfinder-form/petfinder-form.component.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  ngOnInit(): void {
+    this.petfinder.getTypes().pipe(
+      take(1)
+    ).subscribe(res => {
+      this.types = res.types;
+    });
+  }
+
+  getBreeds() {
+    this.petfinder.getBreeds(this.types[this.selectedType]._links.breeds.href)
```
Comment (swhook52) [inline src/app/petfinder/components/petfinder-form/petfinder-form.component.ts:64]: you should manage the lifecycle of this observable with a take(1) or other means of unsubscribing.

File: src/app/petfinder/components/petfinder-form/petfinder-form.component.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        this.nextPageLink = res.pagination._links.next.href;
+      }
+      if (this.currentPage !== 1) {
+        this.previousPageLink = res.pagination._links.previous.href;
+      }
+      this.isLoading = false;
+    });
+  }
+
+  replaceChars(str: string | null): string {
```
Comment (swhook52) [inline src/app/petfinder/components/petfinder-form/petfinder-form.component.ts:148]: This is a good candidate to unit test.

File: src/app/petfinder/petfinder-service/petfinder.service.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+import { Token, Types, Breeds, Parameters, Pets } from './models';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class PetfinderService implements OnDestroy {
+
+  private apiKey = 'BMftsF57yG0s4ZSeDgOqp67N63b5KBZpWSJXOyx6MhTs7l4Ik8';
+  private apiSecret = 'ykUUhf6okEPpaIGZBLB4BgsXXfkfQTQxL6iJzu4O';
+  private baseUrl = 'https://api.petfinder.com'
```
Comment (swhook52) [inline src/app/petfinder/petfinder-service/petfinder.service.ts:14]: These configurations should go in the environment config file at the root and be brought into this service.

**Review by swhook52 — APPROVED** [review]

**Pull request merged**